## 01. 큰 DataFrame 살펴보기

In [2]:
import pandas as pd

In [2]:
laptops_df = pd.read_csv('data/laptops.csv')

#### column

In [7]:
laptops_df.columns

Index(['brand', 'model', 'ram', 'hd_type', 'hd_size', 'screen_size', 'price',
       'processor_brand', 'processor_model', 'clock_speed',
       'graphic_card_brand', 'graphic_card_size', 'os', 'weight', 'comments'],
      dtype='object')

In [8]:
laptops_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 167 entries, 0 to 166
Data columns (total 15 columns):
brand                 167 non-null object
model                 167 non-null object
ram                   167 non-null int64
hd_type               167 non-null object
hd_size               167 non-null int64
screen_size           167 non-null float64
price                 167 non-null int64
processor_brand       167 non-null object
processor_model       167 non-null object
clock_speed           166 non-null float64
graphic_card_brand    163 non-null object
graphic_card_size     81 non-null float64
os                    167 non-null object
weight                160 non-null float64
comments              55 non-null object
dtypes: float64(4), int64(3), object(8)
memory usage: 19.7+ KB


row가 167개 있어야 되는데, 모두 167개가 아닌것 --> 값이 비어 있다.

#### sort

In [10]:
laptops_df.sort_values(by='price')
# 내림차순 하고 싶으면 ascending=False
# 새로운 데이터프레임이 아닌 기존에 적용 inplace=True

brand                       model  ram hd_type  hd_size  screen_size  \
148       Acer             Aspire SW3-016     2     ssd       32         10.1   
83        Acer     A315-31CDC UN.GNTSI.001    2     ssd      500         15.6   
108       Acer  Aspire ES-15 NX.GKYSI.010     4     hdd      500         15.6   
100       Acer    A315-31-P4CRUN.GNTSI.002    4     hdd      500         15.6   
73        Acer             Aspire ES1-523     4     hdd     1024         15.6   
..         ...                         ...  ...     ...      ...          ...   
154  Microsoft      Surface Book CR9-00013    8     ssd      128         13.5   
31        Acer                 Predator 17   16     ssd      256         17.3   
96   Alienware          AW13R3-7000SLV-PUS    8     ssd      256         13.3   
90   Alienware                 15 Notebook   16     hdd     1024         15.6   
5        Apple      MacBook Pro (TouchBar)   16     ssd      512         15.0   

      price processor_brand processor_model  clock_speed graphic_card_brand  \
148   13872           intel      Atom Z8300         1.44              intel   
83    17990           intel         Celeron         1.10              intel   
108   17990             amd         A4-7210         1.80                amd   
100   18990           intel         pentium         1.10              intel   
73    19465             amd         A4-7210         1.80                amd   
..      ...             ...             ...          ...                ...   
154  178799           intel              i5         1.80              intel   
31   178912           intel              i7         2.60             nvidia   
96   190256           intel              i7         3.00             nvidia   
90   199000           intel              i7         2.60             nvidia   
5    226000           intel              i7         2.70              intel   

     graphic_card_size       os  weight  \
148                NaN  windows     1.2   
83                 NaN  windows     2.1   
108                NaN  windows     2.4   
100                NaN  windows     NaN   
73                 NaN    linux     2.4   
..                 ...      ...     ...   
154                NaN  windows     1.5   
31                 NaN  windows     4.2   
96                 6.0  windows     2.6   
90                 8.0  windows     3.5   
5                  2.0      mac     2.5   

                                              comments  
148                                                NaN  
83                                                 NaN  
108                                                NaN  
100                                                NaN  
73                                                 NaN  
..                                                 ...  
154                                                NaN  
31                                 Integrated Graphics  
96   13.3 inch FHD (1920 x 1080) IPS Anti-Glare 300...  
90      Maximum Display Resolution : 1920 x 1080 pixel  
5                                                  NaN  

[167 rows x 15 columns]

## 큰 Series 살펴보기

In [11]:
laptops_df['brand'] # series

0       Dell
1      Apple
2      Apple
3      Apple
4      Apple
       ...  
162     Asus
163     Asus
164     Asus
165     Asus
166     Asus
Name: brand, Length: 167, dtype: object

- 겹치는 값들을 제외하고 보고 싶다면?
- 브랜드별로 몇개씩 보고 싶은지 보고싶다면?
- 브랜드 **시리즈**에 대한 기술적 정보 보고싶다면? **명목형**이어도 가능 !

In [12]:
laptops_df['brand'].describe()

count     167
unique      8
top        HP
freq       55
Name: brand, dtype: object

## 03. 여행지 선정하기1

In [3]:
df = pd.read_csv('data/world_cities.csv', index_col=0)

In [59]:
df.head()

City / Urban area    Country  Population  Land area (in sqKm)
0      Buenos Aires  Argentina    11200000                 2266
1         Melbourne  Australia     3162000                 2080
2            Sydney  Australia     3502000                 1687
3          Brisbane  Australia     1508000                 1603
4             Perth  Australia     1177000                  964

In [90]:
val_count = df['Country'].value_counts()

In [96]:
df['Country'].describe()

count     249
unique     61
top       USA
freq      105
Name: Country, dtype: object

In [95]:
val_count

USA            105
France          15
Brazil          10
Canada           9
Germany          8
              ... 
Sweden           1
Egypt            1
Netherlands      1
Singapore        1
Colombia         1
Name: Country, Length: 61, dtype: int64

1. 총 몇개의 도시와 몇 개의 나라가 있는지?
2. 인구밀도(명/sqKm)가 10000이 넘는 도시?
3. **인구 밀도가 가장 높은 도시?**
4. 4개의 도시만 있는 나라(진행중)

In [97]:
df.head()

City / Urban area    Country  Population  Land area (in sqKm)
0      Buenos Aires  Argentina    11200000                 2266
1         Melbourne  Australia     3162000                 2080
2            Sydney  Australia     3502000                 1687
3          Brisbane  Australia     1508000                 1603
4             Perth  Australia     1177000                  964

In [104]:
country = df['Country'].value_counts()
type(country)

pandas.core.series.Series

In [105]:
country[country==4]

Italy    4
Name: Country, dtype: int64

## 실습과제
2,000명의 코드잇 대학교 학생들이 수강신청을 했습니다.


수강신청에는 다음 3개의 조건이 있습니다.


1. “information technology” 과목은 심화과목이라 1학년은 수강할 수 없습니다.
2. “commerce” 과목은 기초과목이고 많은 학생들이 듣는 수업이라 4학년은 수강할 수 없습니다.
3. 수강생이 5명이 되지 않으면 강의는 폐강되어 수강할 수 없습니다.

기존 DataFrame에 “status”라는 이름의 column을 추가하고, 학생이 수강 가능한 상태이면 “allowed”, 수강 불가능한 상태이면 “not allowed”를 넣어주세요.

3번 시리즈에서 불린값을 찾아서 해당 인덱스를 리스트에 넣고 for을 통해 해결

In [123]:
df = pd.read_csv('data/enrolment_1.csv')

In [124]:
df['status'] = 'allowed'

In [125]:
df.loc[(df['course name'] == 'information technology') & (df['year'] == 1),'status'] = 'not allowed'

In [126]:
df.loc[(df['course name'] == 'commerce') & (df['year'] == 4), 'status'] = 'not allowed'

In [127]:
allowed = df['status']=='allowed'
course_counts = df.loc[allowed, 'course name'].value_counts()
course_counts

arts                         158
science                      124
commerce                     101
english                       56
education                     41
                            ... 
environmental science          1
b.ed                           1
library                        1
school of management           1
home sc. food & nutrition      1
Name: course name, Length: 296, dtype: int64

In [128]:
closed_courses = list(course_counts[course_counts < 5])
# 인덱싱을 안하면 해당 결과값만 나온다.(1,2,3,4,5)
# 인덱싱은 여기서 왼쪽의 과목 이름이 나타날 것.

In [129]:
closed_courses = list(course_counts[course_counts < 5].index)
closed_courses

['chemical',
 'refactoring',
 'fashion communication',
 'computer engg with specialization in cloud computing.',
 'tool & die making',
 'environment',
 'database',
 'information subjects',
 'pathology',
 'electrical engg',
 'engg. & tech.',
 'computer architecture',
 'electronics and instrumentation',
 'korean',
 'agriculture',
 'mechanical engg.',
 'instrumentation engineering',
 'physical education',
 'leather design',
 'mechanical related subjects',
 'electrical and electronics engineering',
 'electros & communicationnic',
 'electronics & communication engineering',
 'applied sc.',
 'data structure',
 'civil engg',
 'electronics and communication',
 'bpt',
 'mech',
 'em',
 'b.com.',
 'politics',
 'science(bzc)',
 'vocational',
 'm.i.l',
 'tax procedure and practice',
 'yes',
 'software computer science',
 'electronics & comm engg',
 'deee',
 'master of computer application',
 'pol-science',
 'bsc mpe',
 'civit',
 'electronics & comm. engg.',
 'administration services',
 'bca',
 'dcp

In [130]:
for course in closed_courses:
    df.loc[df['course name'] == course, 'status'] = 'not allowed'

In [131]:
df['status'].value_counts()

allowed        1448
not allowed     552
Name: status, dtype: int64